semgrep/semgrep/config_resolver.py

```diff
@@ -339,6 +340,68 @@ def _rename_rule_ids(valid_configs: Mapping[str, Sequence[Rule]]) -> None:
                     f"{Config._convert_config_id_to_prefix(config_id)}{rule.id or MISSING_RULE_ID}"
                 )
 
+    @staticmethod
+    def _resolve_patterns_from(valid_configs: Mapping[str, Sequence[Rule]], no_rewrite_rule_ids: bool) -> None:
+        # TODO add support for `pattern`, `pattern-either` etc as top-level keys. This assumes `patterns`
+        # TODO should we be mutating the internal structure of rules or generating a new one? Probably the latter
+        # TODO make sure we update rule.yaml or generate a new one and replace the whole rule
+        # TODO handle circular patterns-from
+
+        def replace_patterns(patterns: Sequence[Dict[str, Any]]) -> Sequence[Dict[str, Any]]:
+            # TODO make copy of patterns and return new_patterns
+            to_replace = {}
+            for idx, item in enumerate(patterns):
+                if 'patterns-from' in item: # TODO handle `patterns-from` not in top level
+                    value = item['patterns-from']
+                    # now resolve them as config objects
+                    config, errors = Config.from_config_list([value]) # TODO validate no errors
+                    if errors:
+                        logger.warning(f"There were errors resolving patterns-from: {errors}")
+                    rules = config.get_rules(no_rewrite_rule_ids)
```

Comment:
Probably fine to resolve in line for now but we probably want something that will precalculate the pattern dependencies and pull them all in at once instead of on demand like this?

Reply:
I implemented that @brendongo